### PR TITLE
Generate dynamic steps

### PR DESCRIPTION
**Pull request by taikiy**
There are dynamic steps that are generated based on the number of input rows to IPA.

Since we are only executing IPA with 10 input rows, it only generates log2(10) (maybe a few more/less because of the optimizations) dynamic steps. We want to generate all possible steps for code generation without actually executing IPA with 1B rows.

This change also suppresses header lines not needed for step colleciton.

### CODE REVIEW

File: scripts/collect_steps.py [hunk 1]
```diff
@@ -16,15 +17,22 @@
...
+BREAKDOWN_KEYS = [8, 33]
 SECURITY_MODEL = ["malicious", "semi-honest"]
 ROOT_STEP_PREFIX = "protocol/alloc::string::String::run-0"
-EXPECTED_HEADER_LINES_COUNT = 4
+
+DEPTH_DYNAMIC_STEPS = "ipa::protocol::attribution::InteractionPatternStep"
+BIT_DYNAMIC_STEPS = [
+    "ipa::protocol::attribution::aggregate_credit::Step::compute_equality_checks",
+    "ipa::protocol::attribution::aggregate_credit::Step::check_times_credit",
+]
```
Comment (taikiy) [inline scripts/collect_steps.py:33]: In the future, we want to do the same thing with a macro annotating dynamic steps and generate code. In order to do that, we need to be able to resolve the step's module path (or source file path) but that functionality is only available in nightly.

File: scripts/collect_steps.py [hunk 2]
```diff
@@ -16,15 +17,22 @@
...
     "--",
     "--num-multi-bits",
-    "1",
+    "3",
 ]
 QUERY_SIZE = 10
 PER_USER_CAP = [1, 3]
 ATTRIBUTION_WINDOW = [0, 86400]
-BREAKDOWN_KEYS = [1, 33]
+BREAKDOWN_KEYS = [8, 33]
```
Comment (akoshelev) [inline scripts/collect_steps.py:25]: is there a difference between 1 and 8?

File: scripts/collect_steps.py [hunk 3]
```diff
@@ -66,14 +67,38 @@ def collect_steps(args):
...
         line = proc.stdout.readline()
         if not line or line == "":
             break
 
         if not line.startswith(ROOT_STEP_PREFIX):
-            header += 1
+            print("Unexpected line: " + line)
+            exit(1)
+
+        # There are protocols in IPA that that will generate log(N) steps where N is the number
```
Comment (akoshelev) [inline scripts/collect_steps.py:79]: maybe this comment belongs to DEPTH_DYNAMIC_STEPS definition

File: scripts/collect_steps.py [hunk 4]
```diff
@@ -16,15 +17,22 @@
...
 ROOT_STEP_PREFIX = "protocol/alloc::string::String::run-0"
-EXPECTED_HEADER_LINES_COUNT = 4
+
+DEPTH_DYNAMIC_STEPS = "ipa::protocol::attribution::InteractionPatternStep"
+BIT_DYNAMIC_STEPS = [
+    "ipa::protocol::attribution::aggregate_credit::Step::compute_equality_checks",
+    "ipa::protocol::attribution::aggregate_credit::Step::check_times_credit",
+]
+MAXIMUM_DEPTH = 32
+MAXIMUM_BIT_LENGTH = 32
```
Comment (akoshelev) [inline scripts/collect_steps.py:35]: while it seems valid to only consider inputs with less than 2B rows, I thought we will and already are decomposing match keys of 40 bits (and looking to use 64 bit match keys in the future)

File: src/telemetry/step_stats.rs [hunk 5]
```diff
@@ -35,6 +35,7 @@ impl CsvExporter for Metrics {
         // then dump them to the provided Write interface
         // TODO: include role dimension. That requires rethinking `Metrics` implementation
         // because it does not allow such breakdown atm.
+        #[cfg(not(feature = "step-trace"))]
```
Comment (akoshelev) [inline src/telemetry/step_stats.rs:38]: maybe we can contain this feature flag to be inside the benchmark only, by configuring test world and metrics by optiionally taking an argument not to print the header

File: scripts/collect_steps.py (hunk 2 above)
Comment (taikiy) [reply scripts/collect_steps.py:25]: There are steps that are dynamically generated based on the number of bits used to represent breakdown keys, and I realized that the coverage wasn't enough with "1". "8" is enough to cover all cases for 1-32 (specialized aggregation protocol). >=33 uses a generic protocol so one is enough to cover all.

File: scripts/collect_steps.py (hunk 4 above)
Comment (taikiy) [reply scripts/collect_steps.py:35]: Bit-dynamic steps are from fields which its bit-length are determined at runtime (e.g. breakdown keys). Match key's bit-length is fixed in the code so the sort protocol generates complete steps.

File: scripts/collect_steps.py (hunk 2 above)
Comment (akoshelev) [reply scripts/collect_steps.py:25]: gotcha. I would suggest to maybe follow up with a comment that explains it inside the code as well. Does not have to be in this PR

**Review by akoshelev — APPROVED** [review]

File: benches/oneshot/ipa.rs [hunk 6]
```diff
@@ -95,13 +95,14 @@ impl Args {
...
 
-    let prep_time = Instant::now();
+    let _prep_time = Instant::now();
     let config = TestWorldConfig {
         gateway_config: GatewayConfig::new(args.active()),
         ..TestWorldConfig::default()
     };
 
     let seed = args.random_seed.unwrap_or_else(|| thread_rng().gen());
+    #[cfg(not(feature = "step-trace"))]
```
[the remark this answers is not in the document]
Comment (taikiy) [reply benches/oneshot/ipa.rs]: Actually, I think this is a useful info to print out in debug mode, too. I made a change together with `print_header` field in Metrics.

File: scripts/collect_steps.py (hunk 2 above)
Comment (taikiy) [reply scripts/collect_steps.py:31]: added comments

File: src/telemetry/step_stats.rs (hunk 5 above)
Comment (taikiy) [reply src/telemetry/step_stats.rs]: done!

**Pull request merged**